src/csharp/Microsoft.Spark/Sql/RowConstructor.cs

```diff
@@ -16,74 +18,94 @@ namespace Microsoft.Spark.Sql
         {
-            return string.Format("{{{0}}}", string.Join(",", _values));
+            _parent = parent;
+            _args = args;
         }
 
         /// <summary>
         /// Used by Unpickler to pass unpickled data for handling.
         /// </summary>
         /// <param name="args">Unpickled data</param>
-        /// <returns>New RowConstructor object capturing unpickled data</returns>
+        /// <returns>New RowConstructor object capturing args data</returns>
         public object construct(object[] args)
         {
-            // Every first call to construct() contains the schema data. When
-            // a new RowConstructor object is returned from this function,
-            // construct() is called on the returned object with the actual
-            // row data.
+            if (s_schemaCache is null)
+            {
+                s_schemaCache = new Dictionary<string, StructType>();
+            }
 
-            // Cache the schema to avoid parsing it for each construct() call.
-            if (s_currentSchema is null)
+            if ((args.Length == 1) && (args[0] is RowConstructor))
             {
-                s_currentSchema = (StructType)DataType.ParseDataType((string)args[0]);
+                args[0] = ((RowConstructor)args[0]).GetRow();
+                return args;
             }
 
-            // Note that on the first call, _values will be set to schema passed in,
-            // but the subsequent construct() call on this object will replace
-            // _values with the actual row data.
-            return new RowConstructor { _values = args, _schema = s_currentSchema };
+            return new RowConstructor(this, args);
         }
 
         /// <summary>
         /// Construct a Row object from unpickled data.
-        /// Resets the cached row schema.
         /// </summary>
         /// <returns>A row object with unpickled data</returns>
         public Row GetRow()
         {
-            // Reset the current schema to ensure that the subsequent construct()
-            // will use the correct schema if rows from different tables are being
-            // unpickled. This means that if pickled data is received per row, the
-            // JSON schema will be parsed for each row. In Spark, rows are batched
-            // such that the schema will be parsed once for a batch.
-            // One optimization is to introduce Reset() to reset the s_currentSchema,
-            // but this has a burden of remembering to call Reset() on the user side,
-            // thus not done.
-            s_currentSchema = null;
+            Debug.Assert(_parent != null);
+
+            for(int i = 0; i < _args.Length; ++i)
+            {
+                if (_args[i] is RowConstructor)
+                {
+                    _args[i] = ((RowConstructor)_args[i]).GetRow();
+                }
+            }
+
+            return new Row(_args, _parent.GetSchema());
+        }
+
+        internal void Reset()
```

Comment:
Comments here as well, when this needs to be called, and why, etc.

Reply:
Added more comments.